### ISSUE TITLE
Type instability when DateTime is inside a function?

### ISSUE THREAD

**Issue by BioTurboNick** [JuliaLang/julia#40362]
```
g() = Dates.DateTime("2005-02-01",Dates.DateFormat("yyyy-mm-dd"))
@code_warntype g()

Variables
  #self#::Core.Const(g)

Body::Any
1 ─ %1 = Dates.DateTime::Core.Const(DateTime)
│   %2 = Dates.DateFormat::Core.Const(DateFormat)
│   %3 = (%2)("yyyy-mm-dd")::DateFormat{Symbol("yyyy-mm-dd"), _A} where _A<:Tuple
│   %4 = (%1)("2005-02-01", %3)::Any
└──      return %4

@code_warntype Dates.DateTime("2005-02-01",Dates.DateFormat("yyyy-mm-dd"))

Variables
  #self#::Core.Const(DateTime)
  dt::String
  df::DateFormat{Symbol("yyyy-mm-dd"), Tuple{Dates.DatePart{'y'}, Dates.Delim{Char, 1}, Dates.DatePart{'m'}, Dates.Delim{Char, 1}, Dates.DatePart{'d'}}}

Body::DateTime
1 ─ %1 = Dates.parse(Dates.DateTime, dt, df)::DateTime
└──      return %1
```

It doesn't seem to affect performance though. Possibly related to this old issue? https://github.com/JuliaLang/julia/issues/13644 Or is it an issue with the `@code_` methods?

**Comment by martinholters** [comment]
Appears fixed on 1.7 (and master).

**Comment by BioTurboNick** [comment]
Confirming this is fixed
